### PR TITLE
`UnitControl`: fix controlled `unit` behavior

### PR DESCRIPTION
**Pull request by ciampo**
## Description

In https://github.com/WordPress/gutenberg/pull/37311#issuecomment-1054591492, @ntsekouras flagged a bug where the `UnitControl` component wouldn't correctly update its unit if the update was coming via the `value` prop (as opposed as from a user interaction).
For reference, these two patterns are also referred to as "controlled" or "uncontrolled" components. `UnitControl` should technically already support this behaviour, but the bug report shows otherwise.

This PR introduces a fix around that, alongside a new Storybook example and a unit test that mimic closely the scenario encountered by Nik when flagging the buggy behavior.

This PR also introduces a new dev dependency, [`@testing-library/user-event`](https://testing-library.com/docs/ecosystem-user-event/), a recent addition to Testing Library for handling interaction in unit tests better.

**Potential follow-ups:**

- Follow-up on https://github.com/WordPress/gutenberg/pull/37311, rewriting the code in a way that doesn't use the deprecated `unit` prop
- Mark the `unit` prop as deprecated also in TypeScript types, and consider adding a soft deprecation warning too
- Rewrite Storybook examples to use controls (and convert to TypeScript)
- Rewrite unit tests to avoid using classnames anduse the `user-event` library (and convert to TypeScript)

## Testing Instructions

- Run `UnitControl` unit tests, and make sure they pass
- Visit the "Controlled" Storybook example for `UnitControl` and make sure it works as expected (changing the unit in one control triggers that same unit change in the other control)

Comment out the fix added by this PR to the component:
- Run unit tests, the recently added test should fail
- Interact with the Storybook example, it should not work as expected

## Screenshots 

Before:

https://user-images.githubusercontent.com/1083581/156259323-075f7d23-1e58-465e-8b53-78759f5ea428.mp4

After:

https://user-images.githubusercontent.com/1083581/156259333-c970c778-0768-4049-aec5-c59d87dd142c.mp4

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

File: package.json
Comment (andrewserong) [inline package.json]: Linting is complaining that this line is not in alphabetical order 😀

**Comment by amustaque97** [comment]
Can we fix `static analysis` stage ` ✖ prefer-alphabetical-devDependencies - node: devDependencies - Your devDependencies are not in alphabetical order. Please move @testing-library/user-event after @testing-library/react.`

Changes LGTM 🚀

**Review by andrewserong — APPROVED** [review]
Nice fix @ciampo, thanks for the thoroughness with the addition of the storybook example and test coverage 👍 — that's all testing well for me, as is standard usage of the component in the block editor.

Aside from the linting issue in `package.json` this looks good to me, so I'll give it the ✅.

I like the addition of `user-event`, too, it's quite a bit easier to work with than `fireEvent` for more complex interactions. I'll just CC @gziolo, though, as he had feedback on another PR I believe that also was looking at adding in another devDependency (from memory, I think it was confidence checking how common it'll be to use the new things we add in as dependencies). In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of `fireEvent` ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

**Comment by ciampo** [comment]
Thank you for the quick review!

I changed the order of the dependencies (somehow it wasn't flagged locally on my machine), and rebased to fix a conflict.

> In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of fireEvent ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

That's what I thought as well — `user-event` makes unit tests even easier to write _and_ read. I'll wait for @gziolo 's blessing before merging, though.

**Comment by gziolo** [comment]
> I like the addition of `user-event`, too, it's quite a bit easier to work with than `fireEvent` for more complex interactions. I'll just CC @gziolo, though, as he had feedback on another PR I believe that also was looking at adding in another devDependency (from memory, I think it was confidence checking how common it'll be to use the new things we add in as dependencies). In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of `fireEvent` ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

There is one thing I don't like about the recommendations they included:

> Most projects have a few use cases for fireEvent, but the majority of the time you should probably use [@testing-library/user-event](https://testing-library.com/docs/ecosystem-user-event).

It too often leads to fragmentation in projects where you have old ways of writing tests, and suddenly there is a better way that impacts the development process because some people might consider it a best practice that everyone should follow. We still have tests written in Enzyme that were never refactored despite the fact that the project is no longer maintained by original creators. 

It looks like there is a growing need for having better testing tools so feel free to land this library as I see there is an agreement that the benefits are bigger than the potential maintenance cost. As far as I can tell, this library depends only on `@testing-library/dom` that is already in use.

For the future, I encourage you to make this type of proposal more transparent. A separate targeted PR with a refactored example test(s), documentation, a list of pros and cons included would give other contributors a chance to leave their feedback.

**Comment by ciampo** [comment]
> For the future, I encourage you to make this type of proposal more transparent. A separate targeted PR with a refactored example test(s), documentation, a list of pros and cons included would give other contributors a chance to leave their feedback.

Hey @gziolo , you definitely make a good point here. I am also happy to remove the dependency, rewrite the tests with `fireEvent`, and open a separate PR which targets the new dependency separately. Let me know if you'd prefer this route instead.

File: packages/components/src/unit-control/test/index.js
Comment (mirka) [inline packages/components/src/unit-control/test/index.js:30]: Non-blocking: Was this story only useful for testing this particular fix or do you think there is value in keeping it around? It's not immediately clear to me whether the main purpose of the story is to demonstrate a particular behavior/implementation, or to provide a debugging playground moving forward for the synced usage. My personal impression is that unit tests are sufficient and better for this kind of behavioral QA, unless you expect frequent manual testing of this "controlled" case in the near future.

**Comment by gziolo** [comment]
> > For the future, I encourage you to make this type of proposal more transparent. A separate targeted PR with a refactored example test(s), documentation, a list of pros and cons included would give other contributors a chance to leave their feedback.
> 
> Hey @gziolo , you definitely make a good point here. I am also happy to remove the dependency, rewrite the tests with `fireEvent`, and open a separate PR which targets the new dependency separately. Let me know if you'd prefer this route instead.

Both are fine. I don't have strong opinions on that at this point since I'm not opposed to the idea after a short investigation I did in regards to this small utilities library 😄

**Comment by ciampo** [comment]
Update: I removed `@testing-library/user-event` from this PR and refactored the unit tests to use `fireEvent` instead.

I'll open a separate PR in the upcoming days about the introduction of the library to Gutenberg, highlighting the pros and showcasing a couple of examples.

I'll merge once CI is 🟢

**Review by mirka — COMMENTED** [review]
Big +1 for adding `user-event` 😃

File: packages/components/src/unit-control/test/index.js
Comment (ciampo) [reply packages/components/src/unit-control/test/index.js:30]: You make a good point. This story is actually very targeted to test the buggy behaviour, and probably doesn't bring much value to Storybook users. I'll go ahead and remove it from this PR.

**Comment by ntsekouras** [comment]
Thanks for the fix @ciampo!

**Pull request merged**
